### ISSUE TITLE
Crash when creating/destroying static __thread variable.

### ISSUE THREAD

**Issue by yzhang-TT** [jemalloc/jemalloc#1276]
We keep seeing crash related to a static thread local variable during thread initialization or exiting.

Attached is the backtrace when creating a new thread and allocating resource for a static thread_local object. There's a similar crash happening when freeing the static thread local object.

Does this ring a bell to anyone? I tried to search in the history but couldn't find any matching issues.
We're still using 4.0.0 version...is this something fixed by later versions?
```
#0  0x00007eff0daffee3 in arena_run_tree_first (rbtree=<optimized out>) at src/arena.c:92
#1  arena_bin_runs_first (bin=<optimized out>) at src/arena.c:1849
#2  arena_bin_nonfull_run_tryget (bin=<optimized out>) at src/arena.c:1879
#3  arena_bin_nonfull_run_get (bin=<optimized out>, arena=0x7eff07bc4600) at src/arena.c:1896
#4  arena_bin_malloc_hard (arena=arena@entry=0x7eff07bc4600, bin=bin@entry=0x7eff07bc4c10) at src/arena.c:1950
#5  0x00007eff0db01abf in je_arena_tcache_fill_small (arena=0x7eff07bc4600, tbin=tbin@entry=0x7eff03962060, binind=2, prof_accumbytes=prof_accumbytes@entry=0) at src/arena.c:2008
#6  0x00007eff0db35e44 in je_tcache_alloc_small_hard (tsd=tsd@entry=0x7efef98f6488, arena=<optimized out>, tcache=tcache@entry=0x7eff03962000, tbin=tbin@entry=0x7eff03962060, binind=<optimized out>) at src/tcache.c:79
#7  0x00007eff0daf2dba in je_tcache_alloc_small (zero=false, size=<optimized out>, tcache=0x7eff03962000, arena=<optimized out>, tsd=0x7efef98f6488) at include/jemalloc/internal/tcache.h:280
#8  je_arena_malloc (tcache=0x7eff03962000, zero=false, size=<optimized out>, arena=<optimized out>, tsd=0x7efef98f6488) at include/jemalloc/internal/arena.h:1097
#9  je_iallocztm (arena=0x0, is_metadata=false, tcache=0x7eff03962000, zero=false, size=<optimized out>, tsd=0x7efef98f6488) at include/jemalloc/internal/jemalloc_internal.h:887
#10 je_imalloc (size=<optimized out>, tsd=0x7efef98f6488) at include/jemalloc/internal/jemalloc_internal.h:906
#11 imalloc_body (usize=<synthetic pointer>, tsd=<synthetic pointer>, size=<optimized out>) at src/jemalloc.c:1395
#12 malloc (size=<optimized out>) at src/jemalloc.c:1410
#13 0x00007eff089516fc in operator new (sz=24) at ../../../../gcc-4.9.1/libstdc++-v3/libsupc++/new_opnt.cc:42
#14 0x00007eff0894f506 in __cxxabiv1::__cxa_thread_atexit at ../../../../gcc-4.9.1/libstdc++-v3/libsupc++/atexit_thread.cc:130
```

**Comment by yzhang-TT** [comment]
Looks like could be something fixed in 4.0.1. Not certain if it's exactly the same issue. https://github.com/jemalloc/jemalloc/commit/45e9f66c280e1ba8bebf7bed387a43bc9e45536d

**Comment by davidtgoldblatt** [comment]
Ah, yep; that looks likely. Handling allocation during bootstrapping is unfortunately quite tricky. Closing for now, feel free to reopen if it starts looking unrelated.
